internal/infrastructure/kubernetes/proxy/resource_provider_test.go

```diff
@@ -1200,148 +1193,93 @@ func TestService(t *testing.T) {
 
-func TestServiceAccount(t *testing.T) {
-	cfg, err := config.New(os.Stdout)
-	require.NoError(t, err)
-	cases := []struct {
-		name                 string
-		infra                *ir.Infra
-		gatewayNamespaceMode bool
-	}{
-		{
-			name:  "default",
-			infra: newTestInfra(),
-		},
-		{
-			name: "with-annotations",
-			infra: newTestInfraWithAnnotations(map[string]string{
-				"anno1": "value1",
-				"anno2": "value2",
-			}),
-		},
-		{
-			name:                 "gateway-namespace-mode",
-			infra:                newTestInfraWithNamespace("ns1"),
-			gatewayNamespaceMode: true,
-		},
-	}
+				r := NewResourceRender(ns, cfg.ControllerNamespace, cfg.DNSDomain, tc.infra.GetProxyInfra(), cfg.EnvoyGateway)
 
-	for _, tc := range cases {
-		t.Run(tc.name, func(t *testing.T) {
-			ns := cfg.ControllerNamespace
-			if tc.gatewayNamespaceMode {
-				deployType := egv1a1.KubernetesDeployModeType(egv1a1.KubernetesDeployModeTypeGatewayNamespace)
-				cfg.EnvoyGateway.Provider = &egv1a1.EnvoyGatewayProvider{
-					Type: egv1a1.ProviderTypeKubernetes,
-					Kubernetes: &egv1a1.EnvoyGatewayKubernetesProvider{
-						Deploy: &egv1a1.KubernetesDeployMode{
-							Type: &deployType,
+				for _, step := range []renderStep{
+					{
+						render: func(r *ResourceRender) (client.Object, error) {
+							return r.ServiceAccount()
+						},
+						resourceType: "serviceaccount",
+						obj:          &corev1.ServiceAccount{},
+					},
+					{
+						render: func(r *ResourceRender) (client.Object, error) {
+							return r.ConfigMap("")
 						},
+						resourceType: "configmap",
+						obj:          &corev1.ConfigMap{},
 					},
+					{
+						render: func(r *ResourceRender) (client.Object, error) {
+							return r.Service()
+						},
+						resourceType: "service",
+						obj:          &corev1.Service{},
+					},
+				} {
+					obj, err := step.render(r)
+					require.NoError(t, err)
+
+					filename := testDataFileName(mode, tc.caseName, step.resourceType)
+					if test.OverrideTestData() {
+						y, err := yaml.Marshal(obj)
+						require.NoError(t, err)
+						require.NoError(t, os.WriteFile(filename, y, 0o600))
+						continue
+					}
+
+					require.NoError(t, loadResource(step.obj, filename))
+					assert.Equal(t, step.obj, obj)
 				}
-				ns = tc.infra.GetProxyInfra().Namespace
-			}
-			r := NewResourceRender(ns, cfg.ControllerNamespace, cfg.DNSDomain, tc.infra.GetProxyInfra(), cfg.EnvoyGateway)
-			sa, err := r.ServiceAccount()
-			require.NoError(t, err)
-
-			if test.OverrideTestData() {
-				saYAML, err := yaml.Marshal(sa)
-				require.NoError(t, err)
-				// nolint: gosec
-				err = os.WriteFile(fmt.Sprintf("testdata/serviceaccount/%s.yaml", tc.name), saYAML, 0o644)
-				require.NoError(t, err)
-				return
-			}
-
-			expected, err := loadServiceAccount(tc.name)
-			require.NoError(t, err)
+			})
+		}
+	}
+}
 
-			assert.Equal(t, expected, sa)
-		})
+func testDataFileName(deployMode egv1a1.KubernetesDeployModeType, tc, resourceType string) string {
```

Comment:
Do we want to sync the gateway mode added here with the test already added for multiple infra proxies?
https://github.com/envoyproxy/gateway/blob/ca8d97ed1ee895fe5649e426f60ce3f38f78715d/internal/infrastructure/kubernetes/proxy/resource_provider_test.go#L1608

Reply:
Either is fine to me.